### ISSUE TITLE
Feature request: Watch mode should rerun affected specs when a file in filesToWatch changes

### ISSUE THREAD

**Issue by seanpoulter** [webdriverio/webdriverio#5668]
**Is your feature request related to a problem? Please describe.**
I'm disappointed when I run in watch mode and I save one of my page objects that is configured in `filesToWatch` but nothing happens. No specs rerun.

<br/>

**Describe the solution you'd like**
I would like to analyze the dependency graph to determine what specs use a file that is saved.

Analyzing the dependency graph of all the specs that are being run is an expensive action but that's OK (to me). It is an acceptable tradeoff compared to finding and saving the spec to trigger the next run.

Analyzing the dependency graph requires knowledge about how the compiler is configured. We don't need to solve this problem. If we can provide sane defaults and allow the user to override the behaviour, it'll be "good enough".

This does not need to be a breaking change. We can allow users to opt-in by adding a new property in the config file. I would like to be able to modify the `chokadir.watch` `'change'` event handler so we can handle `filesToWatch` to watch differently. We will also need to update the `run` method to expect one or more specs to rerun.

I'd propose implementing this in 2 phases:

1. Make it an extensible behaviour
  - Add a new configuration property that is a function that overrides the behaviour of `getFileListener(false)`. I can have a signature like `(configFile, specs, fileThatChanged) => spec[]`
  - Refactor watcher.js to handle more than one `spec` being passed to `run`

2. Provide examples of it working, maybe, as a third-party package?
  - Document how to traverse the dependency graph with Babel
  - Document how to traverse the dependency graph with TypeScript
  - Document the gotchas.

Thoughts on an implementation:

* We want to use a tool to parse the code instead of evaluate it to avoid causing side effects.
* The MVP demo is to parse `import` statements. They have constant module names. This makes it easy to build a depth-first search across a dependency graph from our specs to the file that was saved.
* We/I need to be mindful of CommonJS `require` and dynamic `import()` statements that can be variable. We're all guilty of doing something like ``const env = require(`config/${env}.json`)`` at some point.

<br/>

**Describe alternatives you've considered**

* N/A

<br/>

**Additional context**
* At a minimum, I'm keen to make this an extensible behaviour so I can hack at night and have a slightly better developer experience in my day job. 😉 
* Link back to [gitter.im](https://gitter.im/webdriverio/webdriverio?at=5f1ed055cd7bed0e379420920). Thanks for the reply Christian.
* The flow through [wdio-cli/src/watcher.js](https://github.com/webdriverio/webdriverio/blob/1944ecc93126eea68d6d4e8192ebd6dad11e73c7/packages/wdio-cli/src/watcher.js#L103-L111):
  * If `filesToWatch` is in the `config`, we register a `'change'` event listener for [`chokadir.watch`](https://github.com/webdriverio/webdriverio/blob/1944ecc93126eea68d6d4e8192ebd6dad11e73c7/packages/wdio-cli/src/watcher.js#L40) to call `this.getFileListener` with `passOnFile` set to `false`
  * [`getFileListener`](https://github.com/webdriverio/webdriverio/blob/1944ecc93126eea68d6d4e8192ebd6dad11e73c7/packages/wdio-cli/src/watcher.js#L69) returns a function that calls `this.run({ ...this.args })` with no `spec`
  * [`run `](https://github.com/webdriverio/webdriverio/blob/1944ecc93126eea68d6d4e8192ebd6dad11e73c7/packages/wdio-cli/src/watcher.js#L103) calls `this.getWorkers` using a conditional `pickFn`
* A great video showing an approach (without cycle detection) - https://www.wix.engineering/post/build-your-own-webpack-in-30-minutes

**Comment by jags14385** [comment]
Can I pick this issue up ? Technically , would be my 2nd issue ?

**Comment by christian-bromann** [comment]
@jags14385 please do!

**Comment by seanpoulter** [comment]
I'm available if you need any more info @jags14385. Good luck!

**Comment by jags14385** [comment]
Awesome will keep that in mind. Will work on it over the weekend :)

**Comment by kohlmannj** [comment]
I encountered a bug in which setting the `--spec` command line option failed to re-run test specs when one of the files matching the `filesToWatch` glob patterns changed.

The root causes are these two lines in @wdio/cli's Watcher class (`watcher.ts`):
1. In `this.watch()`, we correctly call `chokidar.watch()` with the `filesToWatch` patterns and add a bound event listener for add and change events which does not include the path to the file ([source](https://github.com/webdriverio/webdriverio/blob/341d306ddea85dc89987a0c11e447f2db3985b93/packages/wdio-cli/src/watcher.ts#L40-L48))
2. The bound event listener calls `this.run()` with argument `this._args`, which includes property `spec` of type `string[]` when `--spec` is set on the command line ([source](https://github.com/webdriverio/webdriverio/blob/341d306ddea85dc89987a0c11e447f2db3985b93/packages/wdio-cli/src/watcher.ts#L78))
3. Since `params.spec` is defined, `this.run()` calls `this.getWorkers()` with a `pickByFn` argument that attempts to see if the worker's `spec` array of strings contains another array of strings `params.spec` ([source](https://github.com/webdriverio/webdriverio/blob/341d306ddea85dc89987a0c11e447f2db3985b93/packages/wdio-cli/src/watcher.ts#L111)). We do this instead of calling `this.getWorkers(undefined)`, which would get us all the workers
4. The value of the local constant `workers` is an empty object (`{}`) since no workers' `spec` arrays of strings included the `spec` array of strings from the command line, so `this.run()` returns early, without re-running any of the specs ([source](https://github.com/webdriverio/webdriverio/blob/341d306ddea85dc89987a0c11e447f2db3985b93/packages/wdio-cli/src/watcher.ts#L114-L119))

I have proposed a fix for this particular bug in #6373.
